lib/astutils.cpp

```diff
@@ -133,9 +133,74 @@ const Token * astIsVariableComparison(const Token *tok, const std::string &comp,
     return ret;
 }
 
+const Token * getVariableInitExpression(const Variable * var)
+{
+    if(!var || !var->declEndToken())
+        return nullptr;
+    if(Token::Match(var->declEndToken(), "; %varid% =", var->declarationId()))
+        return var->declEndToken()->tokAt(2)->astOperand2();
+    return var->declEndToken()->astOperand2();
+}
+
+bool isInLoopCondition(const Token * tok) {
+    return Token::Match(tok->astTop()->previous(), "for|while (");
+}
+
+
+/// This takes a token that refers to a variable and it will return the token
+/// to the expression that the variable is assigned to. If its not valid to
+/// make such substitution then it will return the original token.
+const Token * followVariableExpression(const Token * tok, bool cpp)
+{
+    if(!tok)
+        return tok;
+    // Skip array access
+    if(Token::Match(tok, "%var% ["))
+        return tok;
+    // Skip following variables if it is used in an assignment
+    if(Token::Match(tok->astParent(), "%assign%") || Token::Match(tok->next(), "%assign%"))
+        return tok;
+    const Variable * var = tok->variable();
+    const Token * varTok = getVariableInitExpression(var);
+    if(!varTok)
+        return tok;
+    // Skip array access
+    if(Token::simpleMatch(varTok, "["))
+        return tok;
+    if(var->scope() != tok->scope() && !var->isConst())
```

Comment:
I am not sure why you use `var->scope() != tok->scope()`. Bugs are not found in inner scopes are they? Is that intentional?

Reply:
The problem is that variables can be modified in the outer scope(perhaps in a loop). This test would fail:

```cpp
void f(int b) { 
    int a = 1; 
    while (b) { 
        if ( a != 1){} 
        a++; 
    } 
}
```

I could check if its not modified in the entire scope if its not at the same scope.